lib/sms/sms_submit.c

```diff
@@ -0,0 +1,360 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <logging/log.h>
+#include <stdio.h>
+#include <zephyr.h>
+#include <errno.h>
+#include <modem/at_cmd.h>
+#include <modem/sms.h>
+
+#include "string_conversion.h"
+
+
+LOG_MODULE_DECLARE(sms, CONFIG_SMS_LOG_LEVEL);
+
+/** @brief User Data Header size in octets/bytes. */
+#define SMS_UDH_CONCAT_SIZE_OCTETS 6
+/** @brief User Data Header size in septets. */
+#define SMS_UDH_CONCAT_SIZE_SEPTETS 7
+/** @brief Maximum length of the response for AT commands. */
+#define SMS_AT_RESPONSE_MAX_LEN 256
+/** @brief Buffer size reserved for CMGS AT command.
+ * @details SMS data is maximum of 140 bytes and SMS SUBMIT 24 bytes. This is doubled in
+ * hexadecimal representation an then there is AT command overhead so we'll reserve safe value.
+ */
+#define SMS_AT_CMGS_BUF_SIZE 400
```

Comment:
In one function we allocate this much memory on the stack, which puts a requirement on the caller stack size.
In another function, we allocate this much memory on the heap, which puts a requirement on the heap size.

Can we allocate this buffer statically instead?

Reply:
For the heap case, we can use static allocation but it'll be multiples of CONFIG_SMS_SEND_CONCATENATED_MSG_MAX_CNT, which by default is 2.
If we want to have one static buffer, we would need to do the processing for concatenated message twice. First to find out number of messages needed to send the given text, and then to have each message in the static buffer from where we do sending.

Reply:
As we discussed today, I will remove these allocations and do sending of concatenated message so that I first process through the data to get number of messages required, and then encode again and send each message. We need to encode twice to be able to first determine number of messages that has to be inserted into each message.
This is something I still need to do tomorrow.

Reply:
Sending has been refactored and there is one 160 byte buffer (user_data variable) in the stack and the rest are in static buffers that are reused in receiving side.